### PR TITLE
treewide: use asprintf/strdup for string creations

### PR DESCRIPTION
**Pull request by stoeckmann**
These functions are easier to review than a combination of malloc, strcpy, and strcat.

### CODE REVIEW

File: modules/pam_filter/pam_filter.c [hunk 1]
```diff
@@ -128,92 +128,54 @@ static int process_args(pam_handle_t *pamh
...
 
 	strcpy(levp[0], ARGS_NAME);
-	size = ARGS_OFFSET;
 	for (i=0; i<argc; ++i) {
 	    if (i)
-		levp[0][size++] = ' ';
-	    strcpy(levp[0]+size, argv[i]);
-	    size += strlen(argv[i]);
+		strcat(levp[0], " ");
+	    strcat(levp[0], argv[i]);
```
Comment (BenBE) [inline modules/pam_filter/pam_filter.c:134]: Potentially quadratic runtime behavior (for long strings).

File: modules/pam_filter/pam_filter.c [hunk 2]
```diff
@@ -128,92 +128,54 @@ static int process_args(pam_handle_t *pamh
...
-#define SERVICE_OFFSET    (sizeof(SERVICE_NAME) - 1)
-
 	retval = pam_get_item(pamh, PAM_SERVICE, &tmp);
 	if (retval != PAM_SUCCESS || tmp == NULL) {
 	    pam_syslog(pamh, LOG_CRIT, "service name not found");
-	    if (levp) {
-		free(levp[0]);
-		free(levp);
-	    }
+	    free(levp[0]);
```
Comment (BenBE) [inline modules/pam_filter/pam_filter.c:142]: This still needs the `if (levp)` check.

File: modules/pam_filter/pam_filter.c [hunk 3]
```diff
@@ -128,92 +128,54 @@ static int process_args(pam_handle_t *pamh
...
 
-	levp[1] = malloc(size+1);
-	if (levp[1] == NULL) {
+	if (asprintf(&levp[1], "SERVICE=%s", (const char *) tmp) < 0) {
 	    pam_syslog(pamh, LOG_CRIT, "no memory for service name");
-	    if (levp) {
-		free(levp[0]);
-		free(levp);
-	    }
+	    free(levp[0]);
```
Comment (BenBE) [inline modules/pam_filter/pam_filter.c:149]: Missing `if(levp)` check.

File: modules/pam_filter/pam_filter.c [hunk 4]
```diff
@@ -128,92 +128,54 @@ static int process_args(pam_handle_t *pamh
...
-	if (levp[2] == NULL) {
+	if (asprintf(&levp[2], "USER=%s", user) < 0) {
 	    pam_syslog(pamh, LOG_CRIT, "no memory for user's name");
-	    if (levp) {
-		free(levp[1]);
-		free(levp[0]);
-		free(levp);
-	    }
+	    free(levp[1]);
+	    free(levp[0]);
```
Comment (BenBE) [inline modules/pam_filter/pam_filter.c:163]: Missing `if(levp)` check.

File: modules/pam_filter/pam_filter.c [hunk 5]
```diff
@@ -128,92 +128,54 @@ static int process_args(pam_handle_t *pamh
...
 	    pam_syslog(pamh, LOG_CRIT, "no memory for type");
-	    if (levp) {
-		free(levp[2]);
-		free(levp[1]);
-		free(levp[0]);
-		free(levp);
-	    }
+	    free(levp[2]);
+	    free(levp[1]);
+	    free(levp[0]);
```
Comment (BenBE) [inline modules/pam_filter/pam_filter.c:174]: dito.

File: modules/pam_filter/pam_filter.c (hunk 2 above)
Comment (stoeckmann) [reply modules/pam_filter/pam_filter.c:142]: Impossible for levp to be NULL here. See lines above: https://github.com/linux-pam/linux-pam/blob/b07b6374bc1a5ce11d1a767f7d5cbcc53bd15480/modules/pam_filter/pam_filter.c#L123

File: modules/pam_filter/pam_filter.c (hunk 3 above)
Comment (stoeckmann) [reply modules/pam_filter/pam_filter.c:149]: See above

File: modules/pam_filter/pam_filter.c (hunk 4 above)
Comment (stoeckmann) [reply modules/pam_filter/pam_filter.c:163]: See above

File: modules/pam_filter/pam_filter.c (hunk 5 above)
Comment (stoeckmann) [reply modules/pam_filter/pam_filter.c:174]: See above

File: modules/pam_filter/pam_filter.c (hunk 1 above)
Comment (stoeckmann) [reply modules/pam_filter/pam_filter.c]: Adjusted to use stpcpy (POSIX 2008). Alternatively the commit could simply be removed.

File: modules/pam_xauth/pam_xauth.c [hunk 6]
```diff
@@ -598,10 +584,9 @@ pam_sm_open_session (pam_handle_t *pamh, int flags UNUSED,
 						      getuid(), getgid(),
 						      xauth, (const char *[]) {
 						      xauth, "-f", cookiefile,
-						      "nlist", t, NULL});
+							      "nlist", t, NULL});
```
Comment (BenBE) [inline modules/pam_xauth/pam_xauth.c:587]: Indentation change?

Comment (stoeckmann) [reply modules/pam_xauth/pam_xauth.c]: Fixed

**Review by ldv-alt — APPROVED** [review]

**Pull request merged**